docs/proposals/shoot-control-plane-migration/README.md

```diff
@@ -0,0 +1,228 @@
+
+* Shoot Machine-controller-manager
+
+  The possibility of two machine-control-managers (in source and destination seed clusters) managing the same set of machines should be avoided.
+
+* Shoot Etcd StatefulSet
+
+  The possibility of two etcd statefulsets (in source and destination seed clusters) backing up data to the same object store should be avoided.
+
+* Other control-plane components
+
+Such as logging and tracing or even components that might be introduced in the future.
+
+#### Landscape Diagram
+
+![Landscape diagram](landscape-diagram.png)
+
+#### Components of the solution
+
+##### Lease Service
+
+The Lease Service will act as the central point of truth for all the seed clusters about which seed owns a particular shoot.
+
+* It needs to be accessible from all the seed clusters among which the shoot control-plane can be moved.
+* It needs to be highly available to reduce chances of false positives/negatives.
+* Given the particular scenario of garden/seed/shoot architecture, shoot resource in the garden cluster is already the true single point of truth. The Lease Service would be just a secondary source of the truth. Hence, there is no need for any explicit mechanism for lease renewal or leader promotion. That can be done automatically based on the shoot resource.
+
+###### Lease data structure for a shoot
+
+* Owning seed cluster
+* Lease Period
+  * Lease start timestamp
+  * Lease end timestamp or TTL
+
+DNS seems well suited to address this problem of lease service. We can use DNS records with short TTL to
+publish/broadcast the owning seed for any particular shoot. DNS is highly available and the short TTL can act like
+the lease. Garden shoot-controller or some additional controller can populate or renew the DNS records based on the 
+shoot resource records in the garden cluster.
+
+##### Shoot Control-plane Co-ordinator
+
+The Shoot Control-plane Co-ordinator is a pod (actually, a deployment) deployed as a part of each of the shoot control-plane in the seed clusters.
+
+It has the responsibility to monitor whether the current seed cluster where it is deployed has the lease for the given shoot using any of the mechanisms chosen [above](#advertising-current-seed-cluster-lease-for-a-shoot-cluster) and take appropriate action to activate/passivate the relevant shoot control-plane components such as machine-controller-manager, etcd statefulset, VPN etc.
+
+It has two states.
+
+1. Active
+
+In the active state, the co-ordinator is able to reach the lease service and knows that the current seed is the owner for the given shoot according to the lease service.
+
+In such a scenario, the co-ordinator makes sure the shoot control-plane components such as machine-controller-manager, etcd statefulset, VPN etc. are scaled up and active.
+
+2. Passive
+
+In the passive state, the co-ordinator is either not able reach the lease service or knows that the current seed is not the owner of the given shoot according to the lease service.
```

Comment:
I am not sure whether we should really also act, if we can't reach the lease service. In our case that's most likely DNS (as discussed), which should always be available, but we can always have a bug or networking misconfiguration.

If we don't do it and two control planes really run in parallel, what's the worst case problems and can we mitigate those?